### DISCUSSION TITLE
networks.yml: services_ip should not be part of networks dictionary

### DISCUSSION THREAD

**Discussion by jpm38** [bluebanquise/bluebanquise#460] [Ideas] [answered]
We are using bluebanquise in the context of our project and we rely on a dynamic inventory plugin based on a database to fill all the networking inventory.
On the other hand, the service IPs are coming from another source of information (text) expressed as FQDN.
as a example:
  services_ip:
    pxe_ip: tftp.region2.svc.kube.local
    pxe_http_ip: http.region2.svc.kube.local
    dns_ip: 10.50.200.10
    time_ip: ntp.region2.svc.kube.local
    repository_ip: http.region2.svc.kube.local

As the ansible merge behavior feature is deprecated, we can not rely on it the merge the two parts of dictionary.

A suggestion would be to split these two set of information like following:

networks:
  net2-1:
    subnet: 10.52.0.0
    prefix: 16
    netmask: 255.255.0.0
    broadcast: 10.52.255.255
    gateway: 10.52.255.254
    is_in_dhcp: true
    is_in_dns: true

nework_services:
  net2-1:
    pxe_ip: tftp.region2.svc.kube.local
    pxe_http_ip: http.region2.svc.kube.local
    dns_ip: 10.50.200.10
    time_ip: ntp.region2.svc.kube.local
    repository_ip: http.region2.svc.kube.local

**Comment by btravouillon** [comment]
Hi,

This would introduce an important breaking change for all the installed user base. We already made a lot of breaking changes for 1.3 that were related to the deprecation of the `hash_behavior=merge` and some performance improvement, I'm not keen to do the same for 1.4. To be honest, I even believe we should have released 2.0 instead of 1.3 because of those changes in the BB API.

Can't you consider adding the service IP addresses to your database backend?

**Reply by jpm38** [reply]
I understand that this would introduce a breaking change, but IMHO, from a design point of view, we should separate the network definition from the services. 
In addition, our database backend is designed to provided only hardware and IPAM information and we do not plan to add any logical information.
As I said previously, our services are FQDN and not IP addresses, and they cannot be stored in such a database.

**Comment by oxedions** [comment, marked as answer]
Hello,

A proposal:

Create a file called `combined_networks.yml` in `internals/group_vars/all/`.

Fill it with:
```
networks: "{{ networks_source_a | combine(networks_source_b , recursive=True, list_merge='append_rp') }}"
```

Then try combining source_a and source_b this way 💡 

Note: you need Ansible 2.10+ to do that. I used 2.11 dev0 for my tests.

**Reply by btravouillon** [reply]
I would not recommend to create custom variables in `internals`, this inventory directory being dedicated to BlueBanquise internals. However it is still possible to do this in `inventory` if you ensure Ansible will evaluate `networks_source_a` and `networks_source_b` **before** `networks`.

**Reply by jpm38** [reply]
We will try the proposal using the combine option and let you know.

**Reply by oxedions** [reply]
I agree to avoid putting this into internals, it was for the example.

**Comment by jpm38** [comment]
We've just tested the combine in ansible 2.9 and it works fine.
Thanks !

**Comment by oxedions** [comment]
Cool ! Thanks for the report 😊
